crates/types/src/partition_table.rs

```diff
@@ -282,6 +379,33 @@ impl PartitionTableBuilder {
         }
     }
 
+    pub fn for_each<F>(&mut self, mut modify: F)
+    where
+        F: FnMut(&PartitionId, &mut TargetPartitionReplicationState),
+    {
+        for (partition_id, partition) in self.inner.partitions.iter_mut() {
+            let mut replica_group_builder = partition.replica_group.clone().into();
+            modify(partition_id, &mut replica_group_builder);
+            let replica_group = ReplicaGroup::from(replica_group_builder);
+
+            // auto detect change in replica group either by
+            // change in length or if items at same index are not
+            // equal anymore.
+            // we can't do `left != right` since IndexSet(s) are equal
+            // if they have the same items regardless of the order.
+            if partition.replica_group.len() != replica_group.len()
+                || partition
+                    .replica_group
+                    .iter()
+                    .zip(replica_group.iter())
+                    .any(|(l, r)| *l.0 != *r.0)
+            {
+                self.modified = true;
+                partition.replica_group = replica_group;
+            }
```

Comment:
Should we add a test that ensures that two differently ordered `replica_groups` are detected to be different? What should be the default behavior if I compare two `ReplicaGroups`? If it should be the same, then we probably need to adjust the `PartialEq` implementation.